### ISSUE TITLE
Consider adding generated sources into OpenSSL release or source tree

### ISSUE THREAD

**Issue by nazar-pc** [openssl/openssl#16188]
## Background
I'm working on [OpenSSL wrap for Meson build system](https://github.com/mesonbuild/wrapdb/pull/113) and there are things that could be done upstream that would help massively.

NOTE: I'm not asking to add alternative build system to OpenSSL here, just to facilitate its support with generally useful changes.

The implementation approach I took actually comes from Node.js (and abuses Node.js implementation): https://github.com/nodejs/node/tree/v16.6.0/deps/openssl

The basic idea there is to generate everything into header and assembly files upfront, such that it is possible to build those files later with any build system (GYP in case of Node.js or Meson in my case).

The biggest issue here is Perl scripts that generate a bunch of platform-specific assembly files. Since those files are not a part of upstream release, they need to be bundled with Meson wrap and thus raise security-related questions.

## Request
It would make things much easier if OpenSSL release tarball (or even source tree) included pre-generated files in a similar way to what Node.js does, such that there is no need to rely on external generators and no need to verify whether generated sources correspond to the upstream release. At least ASM files would be nice to have as the first step.

## Impact
I think this will simplify not only Meson integration I'm working on, but also CMake integration that I know others support and will generally reduce the barrier for OpenSSL usage for many projects.

**Comment by eli-schwartz** [comment]
Note that as statically defined source code which is machine generated, it's fairly common for projects to ship this sort of thing in their `make dist` equivalents, e.g. by specifying the files as EXTRA_DIST in autotools. Same with prebuilt manual pages. In fact, autotools also internally does this to ship the generated build system produced from *.am and configure.ac

Doing so will, as a general rule of thumb, save work by the consumer and in most cases reduce build dependencies in one way or another.

In this case it allows the python-based `meson` build system to avoid *also* having a dependency on perl in the compilation rules.

**Comment by beldmit** [comment]
Do you imply the asm generation for all the supported platforms?

**Comment by nazar-pc** [comment]
Yes, all supported platforms.

In the initial Meson wrap linked above following were included (and Meson generally doesn't support all platforms OpenSSL does), but I don't see why others shouldn't be generated upstream:
* BSD-x86
* BSD-x86_64
* darwin64-arm64-cc
* darwin64-x86_64-cc
* darwin-i386-cc
* linux32-s390x
* linux64-mips64
* linux64-s390x
* linux-aarch64
* linux-armv4
* linux-elf
* linux-ppc
* linux-ppc64
* linux-ppc64le
* linux-x32
* linux-x86_64
* solaris64-x86_64-gcc
* solaris-x86-gcc
* VC-WIN32
* VC-WIN64-ARM
* VC-WIN64A

**Comment by richsalz** [comment]
It is quite likely that some of them can only be generated on a native host.  Nobody has checked AFAIK.

**Comment by nazar-pc** [comment]
For platforms above + aix-gcc and aix64-gcc it is certainly possible because Node.js already does it. Can't say anything about others as I have not tried it.

Even in case it is not possible to generate sources for some platforms, the number of platforms for which it is possible will be a significant improvement if implemented.

**Comment by levitte** [comment]
Currently, this would have to be in form of pre-generated work files on a per config basis.  Because we don't have any indication of the config in the generated asm file names, we may have overlapping file names but with slightly different contents, depending on flavors used when generating the asm files, so having *all* variants at once is risky if not impossiible with our current file names.

However, it would be perfectly possible to generate them, one config at a time, and capture each in a suitably named tarball, which could then be layed over the usual OpenSSL source tree.

[this is food for thought...]

**Comment by nazar-pc** [comment]
Node.js stores each platform in a separate subdirectory to avoid any conflicts, but only supports one config for each: https://github.com/nodejs/node/tree/v16.6.0/deps/openssl

Is there really any variation in ASM files depending on configuration within the same platform?

**Comment by levitte** [comment]
> Is there really any variation in ASM files depending on configuration within the same platform?

Not necessarily, but just to pick an example `crypto/bn/asm/x86_64-gf2m.pl` will result in the exact same assembler file name for many different platforms.

> Node.js stores each platform in a separate subdirectory to avoid any conflicts, but only supports one config for each: https://github.com/nodejs/node/tree/v16.6.0/deps/openssl

The same could be achieved by creating build directories for each config as subdirectories of the source directory, or parallell to the source directory, and bundle that all together.
After configuring each of them, `make build_all_generated` is the command needed to generate *all* generatable files (that includes the assembler stuff)

**Comment by nhorman** [comment]
@nazar-pc given that we've arrived at some consensus on https://github.com/openssl/openssl/issues/16812

Does it make sense to keep this around, or would you prefer to make a new issue for asm source generation once you have a PR in place for a new build system, and you know better how you might better know how said future build system can make use of our current perlasm?

**Comment by nazar-pc** [comment]
I'd say this is a sub-task of implementing any second build system in OpenSSL, regardless of whether it is Meson or CMake or something else. And it is something that probably should be submitted as a separate PR. With that context I do not see why this issue should be closed, but let me know if you have a different point of view.

**Comment by nhorman** [comment]
No, keeping it open is fine by me, but if you agree to take on the creation of a Pr in https://github.com/openssl/openssl/issues/16812 , I think it would make sense to mark this as a community issue for you to tackle as well (though this probably will have to have its own set of discussions, as this will have implications to our release process)

**Comment by nazar-pc** [comment]
Community issue is fair. While I don't have the bandwidth to tackle the issue myself, I can try to organize someone to start making progress by upstreaming Meson wrap piece by piece. Adding generated sources will help both Meson wrap downstream and Node.js since they are doing generation themselves and might be interested in contributing upstream.

@RafaelGSS and @lpinca, I see you were reviewing OpenSSL upgrade PRs in Node.js recently, would you or maybe someone you know among Node.js contributors or maintainers be interested in this issue? This should reduce downstream maintenance burden for you folks and help others in the process as well.

**Comment by nhorman** [comment]
Ok, I'll mark this and #16812 as community issues, please assign them as soon as you are able

**Comment by t8m** [comment]
Closing this as won't fix as this was discussed multiple times with no consensus for putting all generated sources into the source tree. (There was actually firm push to the contrary.)

**Comment by eli-schwartz** [comment]
> Closing this as won't fix as this was discussed multiple times with no consensus for putting all generated sources into the source tree. (There was actually firm push to the contrary.)

Err, just to be clear, was that push to the contrary against putting generated sources in "the source tree" (which as per the first response to this ticket -- by me -- is insane), or was that push to the contrary against including it as part of releasing distributable tarballs?

**Comment by t8m** [comment]
Against putting them in the source tree. However nowadays there is also a push against having the distributable tarballs substantially different from the checkout of the respective tag of the git tree. So effectively the push is against both.

**Comment by eli-schwartz** [comment]
:(

Just to be clear, the "Jia Tan" hack was a social engineering problem, not a "tarballs aren't the same" problem. It is absolutely unfair and unrealistic to pretend that the solution to a malicious upstream maintainer is to avoid generated assets. This kind of ridiculous moral panic is instinctive and without any basis in facts, and it has caused and is continuing to cause on an ongoing basis, a great deal of societal harm.

Even if it was, what we would want is idempotent, reproducibly generated assets and therefore idempotent, reproducible tarballs, which is a trivially solvable problem.
